pkg/api/latest/group_register.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package latest
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+var (
+	allGroups = GroupMetaMap{}
+	// Group is a shortcut to allGroups.Group.
+	Group = allGroups.Group
+	// RegisterGroup is a shortcut to allGroups.RegisterGroup.
+	RegisterGroup = allGroups.RegisterGroup
+	// AllPreferredGroupVersions returns the preferred versions of all
+	// registered groups in the form of "group1/version1,group2/version2,..."
+	AllPreferredGroupVersions = allGroups.AllPreferredGroupVersions
+	// IsRegistered is a shortcut to allGroups.IsRegistered.
+	IsRegistered = allGroups.IsRegistered
+)
+
+// GroupMetaMap is a map between group names and their metadata.
+type GroupMetaMap map[string]*GroupMeta
+
+// RegisterGroup registers a group to GroupMetaMap.
+func (g GroupMetaMap) RegisterGroup(groupMeta GroupMeta) error {
+	groupName := groupMeta.GroupVersion.Group
+	_, found := g[groupName]
+	if found {
+		return fmt.Errorf("group %v is already registered", groupName)
+	}
+	if len(groupMeta.GroupVersions) == 0 {
+		return fmt.Errorf("No version is registered for group %v", groupName)
+	}
+	g[groupName] = &groupMeta
+	return nil
+}
+
+// Group returns the metadata of a group if the gruop is registered, otherwise
+// an erorr is returned.
+func (g GroupMetaMap) Group(group string) (GroupMeta, error) {
+	groupMeta, found := g[group]
+	if !found {
+		return GroupMeta{}, fmt.Errorf("no version is registered for group %v", group)
+	}
+	return *groupMeta, nil
+}
+
+// IsRegistered takes a string and determines if it's one of the registered groups
+func (g GroupMetaMap) IsRegistered(group string) bool {
+	_, found := g[group]
+	return found
+}
+
+// AllPreferredGroupVersions returns the preferred versions of all registered
+// groups in the form of "group1/version1,group2/version2,..."
+func (g GroupMetaMap) AllPreferredGroupVersions() string {
+	if len(g) == 0 {
```

Comment:
probably not worth the special-case check... realistically, will we have no groups registered?